logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrator.kt

```diff
@@ -0,0 +1,98 @@
+import com.wire.kalium.logic.data.conversation.MLSConversationRepository
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.id.toApi
+import com.wire.kalium.logic.feature.SelfTeamIdProvider
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.flatMapLeft
+import com.wire.kalium.logic.functional.foldToEitherWhileRight
+import com.wire.kalium.logic.functional.getOrNull
+import com.wire.kalium.logic.kaliumLogger
+import com.wire.kalium.logic.wrapApiRequest
+import com.wire.kalium.network.api.base.authenticated.conversation.ConvProtocol
+import com.wire.kalium.network.api.base.authenticated.conversation.ConversationApi
+import com.wire.kalium.network.api.base.authenticated.conversation.UpdateConversationProtocolResponse
+import kotlinx.coroutines.flow.first
+
+interface MLSMigrator {
+    suspend fun migrateProteusConversations(): Either<CoreFailure, Unit>
+}
+
+class MLSMigratorImpl(
+    private val selfTeamIdProvider: SelfTeamIdProvider,
+    private val conversationRepository: ConversationRepository,
+    private val mlsConversationRepository: MLSConversationRepository,
+    private val conversationApi: ConversationApi
+) : MLSMigrator {
+
+    override suspend fun migrateProteusConversations(): Either<CoreFailure, Unit> =
+        conversationRepository.getConversationList().flatMap {
+            it.first().filter { conversation ->
+                conversation.protocol is Conversation.ProtocolInfo.Proteus &&
+                    conversation.type == Conversation.Type.GROUP &&
+                    conversation.isTeamGroup() &&
+                    conversation.teamId == selfTeamIdProvider().getOrNull()
+            }.foldToEitherWhileRight(Unit) { conversation, _ ->
+                migrate(conversation.id)
+            }
+        }
+
+    private suspend fun migrate(conversationId: ConversationId): Either<CoreFailure, Unit> {
+        kaliumLogger.i("migrating ${conversationId.toLogString()} to mixed")
+        return updateProtocolToMixed(conversationId)
+            .flatMap {
+                establishConversation(conversationId)
+            }.flatMapLeft {
+                kaliumLogger.w("failed to migrate ${conversationId.toLogString()} to mixed: $it")
+                Either.Right(Unit)
+            }
+    }
+
+    private suspend fun updateProtocolToMixed(conversationId: ConversationId) =
+        wrapApiRequest { conversationApi.updateProtocol(conversationId.toApi(), ConvProtocol.MIXED) }
+            .flatMap {
+                if (it is UpdateConversationProtocolResponse.ProtocolUpdated) {
+                    conversationRepository.fetchConversation(conversationId)
+                } else {
+                    Either.Right(Unit)
+                }
+            }
+
+    private suspend fun establishConversation(conversationId: ConversationId) =
```

Comment:
since we updating the db from the api in `updateProtocolToMixed`, shouldn't all the mls-group establishment process happens automatically? after the sync kinda?

Reply:
No, it doesn't happen automatically. What are you referring to?

Reply:
In normal mls conversation, when we sync the users conversations, we establish the conversations based on their state [wait for rejoin or not-established and...] why we can't rely on that auto recover mechanism?